src/erc4626/ERC4626Oracle.sol

```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {IERC20} from "./IERC20.sol";
+import {IERC4626} from "./IERC4626.sol";
+
+contract ERC4626Oracle {
+    error OracleUnsupportedPair();
+
+    IERC4626 public immutable VAULT;
+    address public immutable ASSET;
+
+    constructor(address _vault) {
+        VAULT = IERC4626(_vault);
+        ASSET = VAULT.asset();
+    }
+
+    function valueOf(address base, address quote, uint256 baseAmount) external view returns (uint256 quoteAmount) {
+        if (quote == address(VAULT) && base == ASSET) {
+            return VAULT.convertToShares(baseAmount);
+        } else if (quote == ASSET && base == address(VAULT)) {
+            return VAULT.convertToAssets(baseAmount);
+        } else {
+            revert OracleUnsupportedPair();
+        }
+    }
+
+    function priceOf(address base, address quote) external view returns (uint256 baseQuotePrice) {
+        if (quote == address(VAULT) && base == ASSET) {
+            return VAULT.convertToShares(10 ** IERC20(ASSET).decimals());
```

Comment:
```suggestion
            return (VAULT.convertToShares(ASSET_SCALAR) * 1e18) / VAULT_SCALAR;
```

From the spec, price is the "value of a whole unit of base in quote terms, multiplied by 10**18 and divided by the quote scalar".

In other words, prices are 18-decimal factors as if base and quote would be decimals as well (instead of integers).